skorch/history.py

```diff
@@ -23,17 +23,108 @@ def _filter_none(items):
     return type_(filter(_not_none, items))
 
 
-def _getitem(item, i):
-    """Extract value or values from dicts.
+def _getitem_list_list(items, keys):
+    """Ugly but efficient extraction of multiple values from a list of
+    items.
 
-    Covers the case of a single key or multiple keys. If not found,
-    return placeholders instead.
+    Keys are contained in a list.
 
     """
-    if not isinstance(i, (tuple, list)):
-        return item.get(i, _none)
-    type_ = list if isinstance(item, list) else tuple
-    return type_(item.get(j, _none) for j in i)
+    filtered = []
+    for item in items:
+        row = []
+        do_append = True
+        for key in keys:
+            try:
+                row.append(item[key])
+            except KeyError:
+                do_append = False
+                break
+        if row and do_append:
+            filtered.append(row)
+    return filtered
+
+
+def _getitem_list_tuple(items, keys):
+    """Ugly but efficient extraction of multiple values from a list of
+    items.
+
+    Keys are contained in a tuple.
+
+    """
+    filtered = []
+    for item in items:
+        row = ()
+        do_append = True
+        for key in keys:
+            try:
+                row += (item[key],)
+            except KeyError:
+                do_append = False
+                break
+        if row and do_append:
+            filtered.append(row)
+    return filtered
+
+
+def _getitem_list_str(items, key):
+    filtered = []
+    for item in items:
+        try:
+            filtered.append(item[key])
+        except KeyError:
+            continue
+    return filtered
+
+
+def _getitem_dict_list(item, keys):
+    return [item.get(key, _none) for key in keys]
+
+
+def _getitem_dict_tuple(item, keys):
+    return tuple(item.get(key, _none) for key in keys)
+
+
+def _getitem_dict_str(item, key):
+    return item.get(key, _none)
+
+
+def _get_getitem_method(items, key):
```

Comment:
I think we can take advantage of https://docs.python.org/3/library/functools.html#functools.singledispatch here instead of dispatching ourselves.

Reply:
Yes, I actually used implemented it both ways. After running a profiler, I saw that there was no difference (presumably `singledispatch` does more or less the same work under the hood).

The main reasons why I didn't go for `singledispatch`:

- we technically need multiple dispatch (`key` and `items`)
- it can look a bit daunting (having to add `@fun.register` everywhere, no nice syntax as in, say, Julia)
- it's presumably not so well known, so harder for readers to understand what's happening, whereas currently, this suggestion only uses "basic" Python language features

What would be the advantages of using `singledispatch` in your opinion?

Reply:
I haven't looked too closely at the implementation, but I think one benefit of Python's singledispatch is that better documentation is available. Additionally, the concise naming will very likely help new developers like me find relevant examples/SO questions.

Reply:
> one benefit of Python's singledispatch is that better documentation is available

Do you mean the documentation for `singledispatch` itself? Arguably, the current code doesn't even need consulting documentation as long as you know what `if isinstance(...)` does :) But readability is of course debatable.

However, consider that for this specific case, we dispatch on two arguments, so we would need either nested `singledispatch` or a mix of `singledispatch` and `isinstance`. If we were dispatching only on one argument or if there were a `multidispatch`, I would tend to agree that this approach is superior.

Reply:
Yeah, nevermind. I misunderstood what the comment was referring too. Multiple `isinstance` blocks is a lot clearer.

Reply:
Lets leave it as is. The current implementation is already a net benefit.